pkg/operator/encryption/migration_controller_encryption_test.go

```diff
@@ -0,0 +1,118 @@
+	scenarios := []struct {
+		name                     string
+		initialResources         []runtime.Object
+		encryptionSecretSelector metav1.ListOptions
+		targetNamespace          string
+		targetGRs                map[schema.GroupResource]bool
+		// expectedActions holds actions to be verified in the form of "verb:resource:namespace"
+		expectedActions []string
+		validateFunc    func(ts *testing.T, actions []clientgotesting.Action)
+	}{
+		// scenario 1
+		{
+			name:            "describe me",
+			targetNamespace: "kms",
+			targetGRs: map[schema.GroupResource]bool{
+				schema.GroupResource{Group: "", Resource: "secrets"}: true,
+			},
+			initialResources: []runtime.Object{
+				createDummyKubeAPIPod("kube-apiserver-1", "kms"),
+				createReadEncryptionKeySecretWithRawKey("kms", schema.GroupResource{"", "secrets"}, 1, []byte("71ea7c91419a68fd1224f88d50316b4e")),
+				func() runtime.Object {
+					keys := []apiserverconfigv1.Key{
+						apiserverconfigv1.Key{
+							Name:   "1",
+							Secret: "NzFlYTdjOTE0MTlhNjhmZDEyMjRmODhkNTAzMTZiNGU=",
+						},
+					}
+					ec := createEncryptionCfgSecretWithWriteKeys(t, keys, "kms", "1", "secrets")
+					return ec
+				}(),
+			},
+			expectedActions: []string{"list:pods:kms", "list:secrets:openshift-config-managed"},
+		},
+	}
+
+	for _, scenario := range scenarios {
+		t.Run(scenario.name, func(t *testing.T) {
+			// setup
+			fakeOperatorClient := v1helpers.NewFakeStaticPodOperatorClient(
+				&operatorv1.StaticPodOperatorSpec{
+					OperatorSpec: operatorv1.OperatorSpec{
+						ManagementState: operatorv1.Managed,
+					},
+				},
+				&operatorv1.StaticPodOperatorStatus{
+					OperatorStatus: operatorv1.OperatorStatus{
+						// we need to set up proper conditions before the test starts because
+						// the controller calls UpdateStatus which calls UpdateOperatorStatus method which is unsupported (fake client) and throws an exception
+						Conditions: []operatorv1.OperatorCondition{
+							operatorv1.OperatorCondition{
+								Type:   "EncryptionStateControllerDegraded",
+								Status: "False",
+							},
+						},
+					},
+				},
+				nil,
+				nil,
+			)
+			fakeKubeClient := fake.NewSimpleClientset(scenario.initialResources...)
+			eventRecorder := events.NewRecorder(fakeKubeClient.CoreV1().Events(scenario.targetNamespace), "test-encryptionKeyController", &corev1.ObjectReference{})
+			// we pass "openshift-config-managed" and $targetNamespace ns because the controller creates an informer for secrets in that namespace.
+			// note that the informer factory is not used in the test - it's only needed to create the controller
+			kubeInformers := v1helpers.NewKubeInformersForNamespaces(fakeKubeClient, "openshift-config-managed", scenario.targetNamespace)
+			fakeSecretClient := fakeKubeClient.CoreV1()
+			fakePodClient := fakeKubeClient.CoreV1().Pods(scenario.targetNamespace)
+			scheme := runtime.NewScheme()
+			fake.AddToScheme(scheme)
```

Comment:
[listing resources](https://github.com/enj/cluster-kube-apiserver-operator/blob/enj/f/enc_config/pkg/operator/encryption/migration_controller_encryption.go#L184) gives me `item[0]: can't assign or convert v1.Secret into unstructured.Unstructured`, is this the correct way of setting up a fake dynamic client?

Reply:
resolved!

Reply:
we need a scheme for the dynamic client so that it can properly discover resources.